test/boost/schema_loader_test.cc

```diff
@@ -96,6 +96,19 @@ SEASTAR_THREAD_TEST_CASE(test_udts) {
     ).get().size(), 2);
 }
 
+SEASTAR_THREAD_TEST_CASE(test_udts_with_dependencies) {
+    db::config dbcfg;
+    BOOST_REQUIRE_EQUAL(tools::load_schemas(
+                dbcfg,
+                "CREATE KEYSPACE ks WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}; "
+                "CREATE TYPE ks.type1 (f1 int); "
+                "CREATE TYPE ks.type2 (f1 int); "
+                "CREATE TYPE ks.type3 (f1 frozen<ks.type1>); "
+                "CREATE TYPE ks.type4 (f1 frozen<ks.type2>, f2 frozen<ks.type3>); "
```

Comment:
The order of the UDTs in this test is uninteresting, i.e. they already are in a valid topological order. Perhaps that's why the test passes despite the bug pointed out in https://github.com/scylladb/scylladb/pull/18302#discussion_r1595463088?

In any case, I suggest adding a bit more types, making a longer chain of them and shuffling them. Also, keep in mind that if you make sure to order the types alphabetically in the `std::map` then your test should account for that. For example, if you have UDTs `a`, `b`, `c`, `d`, `e` then a dependency chain `c` -> `b` -> `d` -> `a` -> `e` is more interesting than a chain `a` -> `b` -> `c` -> `d` -> `e`, even if the order of the UDTs is shuffled and not alphabetical.

Reply:
Wait, this is the new code:

```c++
        auto sorted = co_await utils::topological_sort(all_definitions, adjacency);
        std::vector<user_type> created;
        created.reserve(sorted.size());

        for (auto* e : sorted) {
            created.push_back(e->prepare(ks_name, types));
        }
```

Shouldn't we get a failure from `e->prepare` in the existing test? Shouldn't preparing a UDT that depends on other UDTs require that dependents are put into `types`?

Reply:
> The order of the UDTs in this test is uninteresting, i.e. they already are in a valid topological order.

Gah, what am I saying. I suppose that we are just plainly executing those statements in order. They must be sorted topologically or executing them will fail.

Reply:
> Shouldn't we get a failure from `e->prepare` in the existing test? Shouldn't preparing a UDT that depends on other UDTs require that dependents are put into `types`?

`types` is initialized with all pre-existing types. Over the course of the function it is updated with new instances of the types which were ALTERED (or depend on one that was) or newly created.